modules/bwa/aln/main.nf

```diff
@@ -0,0 +1,67 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+options        = initOptions(params.options)
+
+process BWA_ALN {
+    tag "$meta.id"
+    label 'process_medium'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
+
+    conda (params.enable_conda ? "bioconda::bwa=0.7.17" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/bwa:0.7.17--h5bf99c6_8"
+    } else {
+        container "quay.io/biocontainers/bwa:0.7.17--h5bf99c6_8"
+    }
+
+    input:
+    tuple val(meta), path(reads)
+    path index
+
+    output:
+    tuple val(meta), path(reads), path("*.sai"), emit: sai
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    if (meta.single_end) {
+        """
+        INDEX=`find -L ./ -name "*.amb" | sed 's/.amb//'`
+
+        bwa aln \\
+            $options.args \\
+            -t $task.cpus \\
+            -f ${prefix}.sai \\
+            \$INDEX \\
+            ${reads}
+
+        echo \$(bwa 2>&1) | sed 's/^.*Version: //; s/Contact:.*\$//' > ${software}.version.txt
+        """
+    } else {
+        """
+        INDEX=`find -L ./ -name "*.amb" | sed 's/.amb//'`
+
+        bwa aln \\
```

Comment:
Hmmm...this is probably going to be quite inefficient running these in serial for paired-end data but I don't have a good alternative suggestion. Unless the input channels are tweaked to provide the `process` one fastq at a time 🤔 and we remove all of the logic here that checks if the sample is SE/PE because the module would just take a channel containing fastq files. What do you think? The workflow run-times would be better for it.

Reply:
I get your point, and I actually also considered that originally. However I ended up going against that for the following reason:

1) Puts more effort/complication for a pipeline developer to write custom channel logic to syncronised them as input to SAMSE/SAMPE (although I guess a subworkflow could be made for this)
2) Very few people use `aln` nowadays anyway because there are better methods for longer reads. `aln` is mostly just used in the realm of palaeogenomics I believe, which has very short reads, and we are used to it being slow anyway.
3) It also would break the 'logic' that is expected by users, that the associated files come together, as nextflow couldn't label the associated output files properly, without breaking typical TSV input structure (i.e. which is F and R, how would the input channels know which is which for labelling without putting each FASTQ on a new line - which is then also difficult for downstream processing)

At least for EAGER there will be loads more runtime optimisation due to other modules being even more atomised, so I think having  one module (using old software!) still being a bit slower isn't so bad.

Reply:
Fair enough 👍🏽